ocaml/xapi/xapi_xenops.ml

```diff
@@ -555,31 +555,37 @@ module MD = struct
       | `pv_in_pvh | `pv | `pvh | `unspecified ->
           false
     in
-    let device_number = Device_number.of_string hvm vbd.API.vBD_userdevice in
+    let device_number =
+      match Device_number.of_string ~hvm vbd.API.vBD_userdevice with
+      | Some dev ->
+          dev
+      | None ->
+          raise
+            Api_errors.(Server_error (invalid_device, [vbd.API.vBD_userdevice]))
+    in
     let open Vbd in
     let ty = vbd.API.vBD_qos_algorithm_type in
     let params = vbd.API.vBD_qos_algorithm_params in
     let qos_class params =
-      if List.mem_assoc "class" params then
-        match List.assoc "class" params with
-        | "highest" ->
-            Highest
-        | "high" ->
-            High
-        | "normal" ->
-            Normal
-        | "low" ->
-            Low
-        | "lowest" ->
-            Lowest
-        | s -> (
-          try Other (int_of_string s)
-          with _ ->
-            warn "Unknown VBD QoS scheduler class (try 'high' 'low' 'normal')" ;
-            Normal
-        )
-      else
-        Normal
+      match List.assoc_opt "class" params with
```

Comment:
These are not related to String.implode, might've been useful to have them in a separate commit.

Reply:
Could you update the commit message with a description of what else to expect in this commit? (e.g. the assoc_opt optimization, and I also see some device number handling refactoring)

Reply:
I've described both changes in the commit. Let me know if it's clarifying enough